api/v1alpha1/thanosruler_types.go

```diff
@@ -0,0 +1,99 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ThanosRulerSpec defines the desired state of ThanosRuler
+type ThanosRulerSpec struct {
+	CommonThanosFields `json:",inline"`
+	// Replicas is the number of Ruler replicas.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Required
+	Replicas int32 `json:"replicas,omitempty"`
+	// By default, the operator will add all discoverable Queriers to the Ruler,
+	// if they have query labels. You can optionally choose to override default
+	// Query selector labels, to select a subset of Queries to query.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={matchLabels:{"operator.thanos.io/query-api": "true"}}
+	QueryLabelSelector *metav1.LabelSelector `json:"queryLabelSelector,omitempty"`
+	// ObjectStorageConfig is the secret that contains the object storage configuration for Ruler to upload blocks.
+	// +kubebuilder:validation:Required
+	ObjectStorageConfig ObjectStorageConfig `json:"defaultObjectStorageConfig,omitempty"`
+	// The operator will mount all ConfigMaps with the label operator.thanos.io/rule-file=true
+	// to the Ruler as rule file. You can optionally choose to override default Rule Config selector labels,
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={matchLabels:{"operator.thanos.io/rule-file": "true"}}
+	RuleConfigSelector *metav1.LabelSelector `json:"ruleConfigSelector,omitempty"`
+	// AlertmanagerURL is the URL of the Alertmanager to which the Ruler will send alerts.
+	// +kubebuilder:validation:Required
+	AlertmanagerURL string `json:"alertmanagerURL,omitempty"`
+	// ExternalLabels set on Ruler TSDB, for query time deduplication.
+	// +kubebuilder:default={rule_replica: "$(NAME)"}
+	// +kubebuilder:validation:Required
+	ExternalLabels ExternalLabels `json:"externalLabels,omitempty"`
+	// Labels to drop before Ruler sends alerts to alertmanager.
+	// +kubebuilder:validation:Optional
+	AlertLabelDrop []string `json:"alertLabelDrop,omitempty"`
+	// Retention is the duration for which the Thanos Rule StatefulSet will retain data.
+	// +kubebuilder:default="2h"
+	// +kubebuilder:validation:Required
+	Retention Duration `json:"retention,omitempty"`
+	// StorageSize is the size of the storage to be used by the Thanos Ruler StatefulSet.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$`
+	StorageSize string `json:"storageSize"`
```

Comment:
This doesn't have to be done now, but I wonder if for ruler we should group these somehow into "StatefulConfig" since they should act as a set of configs for stateful ruler and stateless ruler would have its own set

Reply:
Yup when we introduce stateless we can